### PR TITLE
Interface most of Vanilla's builtin AI tasks

### PR DESCRIPTION
**Pull request by liach**
Rebirth of #1012 
Waiting for more input and review.

Todo:
- (Waiting for input)

### CODE REVIEW

File: src/main/java/org/spongepowered/api/entity/ai/task/builtin/VanillaAITask.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+     * Get the mutex bits of an AI task in vanilla. The mutex bit is not
+     * supposed to be used, and might not exist in some different
+     * implementations. See {@link AITask#canRunConcurrentWith(AITask)}, which
+     * achieves the function of mutex bits.
+     *
+     * @deprecated Platform specific, might not be available
+     * @return The mutex bits of this vanilla AI task.
+     */
+    @Deprecated
+    int getMutexBits();
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/entity/ai/task/builtin/VanillaAITask.java]: After sleeping on this, I think the mutex stuff + Vanilla parent interface should move to a separate PR as finding a correct solution to it shouldn't hold up the rest of this PR (which is pretty good).

Since they both target bleeding, breaking the API in the next PR isn't a problem.

File: src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/rangedattacker/RangedAttackAITask.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import org.spongepowered.api.entity.living.RangedAttackingAgent;
+
+public interface RangedAttackAITask extends VanillaAITask<RangedAttackingAgent> {
+
+    //maxRangedAttackTime
+    int getMaxCoolDown();
+
+    RangedAttackAITask setMaxCoolDown(int maxCoolDown);
+
+    //field_96561_g
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/rangedattacker/RangedAttackAITask.java]: These do not belong in the API.

File: src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/movetolocation/MoveToBlockAITask.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+import org.spongepowered.api.entity.ai.task.builtin.VanillaAITask;
+import org.spongepowered.api.entity.ai.task.builtin.VanillaAITaskBuilder;
+import org.spongepowered.api.entity.living.Creature;
+import org.spongepowered.api.world.Location;
+import org.spongepowered.api.world.World;
+
+import java.util.function.Predicate;
+
+public interface MoveToBlockAITask<O extends Creature, A extends MoveToBlockAITask<O, A>> extends VanillaAITask<O> {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/movetolocation/MoveToBlockAITask.java]: Which Vanilla class is this representing...hopefully they aren't silly enough to do the pathfinder in a task...

File: src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/movetolocation/MoveToLivingAITask.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+import java.util.Optional;
+
+public interface MoveToLivingAITask extends VanillaAITask<Creature> {
+
+    /**
+     * Supposed to be stored in {@link com.flowpowered.math.vector.Vector3d}
+     * @return The destination of the task
+     */
+    Location<World> getDestinationAsLocation();
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/movetolocation/MoveToLivingAITask.java]: Is there a reason for the Location usage? Vector3d seems appropriate.

File: src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/movetolocation/MoveToLocationAITask.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+import org.spongepowered.api.world.Location;
+import org.spongepowered.api.world.World;
+
+public interface MoveToLocationAITask extends VanillaAITask<Creature> {
+
+    /**
+     * Supposed to be stored in {@link com.flowpowered.math.vector.Vector3d}
+     * @return The destination of the task
+     */
+    Location<World> getDestination();
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/movetolocation/MoveToLocationAITask.java]: See above.

File: src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/tameable/AttackAfterOwnerAITask.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+    Optional<Living> getEnemy();
+
+    AttackAfterOwnerAITask setEnemy(Living enemy);
+
+    int getEnemyLastAttackTime();
+
+    AttackAfterOwnerAITask setEnemyLastAttackTime(int enemyLastAttackTime);
+
+    interface Builder extends TargetAITask.Builder<Tameable, AttackAfterOwnerAITask, Builder> {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/tameable/AttackAfterOwnerAITask.java]: You are missing builder methods for this particular interface. This is true for many other of your added builders.

File: src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/target/RevengeAITask.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+import java.util.Set;
+
+public interface RevengeAITask extends TargetAITask<RevengeAITask, Creature> {
+
+    boolean hasGroupAnger();
+
+    RevengeAITask setGroupAnger(boolean groupAnger);
+
+    ImmutableSet<Class<?>> getTargetTypes();
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/target/RevengeAITask.java]: Fairly sure you meant ```? extends Living/Entity``` here.

File: src/main/java/org/spongepowered/api/entity/living/golem/IronGolem.java [hunk 8]
```diff
@@ -36,4 +36,8 @@
         return getValue(Keys.PLAYER_CREATED).get();
     }
 
+    default Value<Boolean> isHoldingFlower() {
+        return getValue(Keys.IRON_GOLEM_HOLDING_FLOWER).get();
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/entity/living/golem/IronGolem.java]: Is this value always present on an IronGolem?

File: src/main/java/org/spongepowered/api/entity/living/player/User.java [hunk 9]
```diff
@@ -32,14 +31,17 @@
...
+import org.spongepowered.api.profile.GameProfile;
 import org.spongepowered.api.service.permission.Subject;
 import org.spongepowered.api.util.Identifiable;
 
 import java.util.Optional;
 
 /**
- * A User is the data usually associated with a Player that is persisted across server restarts.
- * This is in contrast to Player which represents the ingame entity associated with an online User.
+ * A User is the data usually associated with a Player that is persisted across
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/entity/living/player/User.java]: This doesn't belong in this PR. Same as the doc changes in Living.

File: src/main/java/org/spongepowered/api/event/entity/ai/AITaskEvent.java [hunk 10]
```diff
@@ -77,6 +77,14 @@
     }
 
     /**
+     * Fired when an {@link AITask} is updated every
+     */
+    @FactoryCodeCheck(errorMessage = "The target entity '%s' is not the owner of the goal '%s'!")
+    interface Tick extends AITaskEvent {
```
Comment (Zidane) [inline src/main/java/org/spongepowered/api/event/entity/ai/AITaskEvent.java]: Not too confident on this event....because of the incredible amount of spam. Put this into another PR please.

**Comment by Zidane** [comment]
Please forgive me @liach ...but I'll need you to document -every- -single- -interface- with what it is, what it does, and -every- -single- -method- which what they mean and what they do.

I'm aware its massive but it is my hope to make this one of the bigger great additions to API 4.0.

File: src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/tameable/AttackAfterOwnerAITask.java (hunk 6 above)
Comment (liach) [inline src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/tameable/AttackAfterOwnerAITask.java]: So do I remove these builders or what?

File: src/main/java/org/spongepowered/api/entity/living/player/User.java (hunk 9 above)
Comment (liach) [inline src/main/java/org/spongepowered/api/entity/living/player/User.java]: Fixing checkstyle - javadoc lines longer than 80

File: src/main/java/org/spongepowered/api/event/entity/ai/AITaskEvent.java (hunk 10 above)
Comment (liach) [inline src/main/java/org/spongepowered/api/event/entity/ai/AITaskEvent.java]: Ok.

File: src/main/java/org/spongepowered/api/entity/living/player/User.java (hunk 9 above)
Comment (Deamon5550) [inline src/main/java/org/spongepowered/api/entity/living/player/User.java]: Thats not the point, this is a PR for changes to AI not fixing checkstyle. Keep it focused and remove this change.

**Comment by liach** [comment]
Accepted, but you need to wait a long time for these changes... Those are a lot. And another note: I did not add AIs that are inner classes of the entity classes, for example, `EntityEnderman$AIPlaceBlock`.

File: src/main/java/org/spongepowered/api/entity/living/golem/IronGolem.java (hunk 8 above)
Comment (liach) [inline src/main/java/org/spongepowered/api/entity/living/golem/IronGolem.java]: This is a field in iron golem's vanilla code which stores this value, though it has no usage in the class itself, as it only has getters and setters.

File: src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/tameable/AttackAfterOwnerAITask.java (hunk 6 above)
Comment (Zidane) [inline src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/tameable/AttackAfterOwnerAITask.java]: I meant you are missing ```enemy(Living enemy)``` and such on many of these.

**Comment by Zidane** [comment]
@liach 

While I'd like to interface the inner ones, its not a priority for me at this time. Lets keep focused to the ai package.

**Comment by Zidane** [comment]
@liach 

Gave this a more fitting title.

File: src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/rangedattacker/RangedAttackAITask.java (hunk 2 above)
Comment (liach) [inline src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/rangedattacker/RangedAttackAITask.java]: Will remove when docs comes.

File: src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/target/RevengeAITask.java (hunk 7 above)
Comment (liach) [inline src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/target/RevengeAITask.java]: Will fix with docs.

File: src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/movetolocation/MoveToLivingAITask.java (hunk 4 above)
Comment (DarkArc) [inline src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/movetolocation/MoveToLivingAITask.java]: @Zidane There could be a mob in the future which utilizes portals/some other mechanic to chase something across worlds.

**Comment by liach** [comment]
By the way, do I need to add blank line at the end of the last member of the class? I saw some classes have such blank lines but some do not.

**Comment by simon816** [comment]
https://docs.spongepowered.org/en/contributing/implementation/codestyle.html
> * Vertical whitespace
  * Place a blank line before the first member of a class, interface, enum, etc. (i.e. after `class Example {`) as well as after the last member

**Comment by liach** [comment]
Well thanks. I saw code like [this](https://github.com/SpongePowered/SpongeAPI/blob/master/src/main/java/org/spongepowered/api/entity/ai/task/AbstractAITask.java#L72) and deleted all blank lines after the last member. I will fix it.

**Comment by simon816** [comment]
it used to be "encouraged but not required", then it changed in https://github.com/SpongePowered/SpongeDocs/commit/f887974b82e871163f0207e10cc8aa56ab4b5a8a#diff-ca033b402e5877a8433cca6da4344ebaL36 (over a year ago), seems some devs haven't caught on

**Comment by liach** [comment]
@Zidane Documentation finished. Any more requests?

File: src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/movetolocation/MoveToLivingAITask.java (hunk 4 above)
Comment (liach) [inline src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/movetolocation/MoveToLivingAITask.java]: Actually not...

**Comment by liach** [comment]
Request `implementation pending` tag. @SpongePowered/developers Reviews?

**Comment by Zidane** [comment]
@liach 

Nice job, I mean seriously. Only request is not putting (MCP Name: something) in the API.

**Comment by ryantheleach** [comment]
Is it possible to make BreakDoorAITask or one similar that can be applied to any block?

One idea I had for a plugin was to modify behaviour of mobs, to target weak / light sources.

**Comment by liach** [comment]
Nope...

**Comment by fullwall** [comment]
Is the intention of the API to be a thin wrapper around how MC functionality would work or provide more high level abstractions? I find the current interfaces to be quite closely linked to MC behaviour (e.g. Predicate<BlockState> being used without a way to affect selection method).

**Comment by Zidane** [comment]
@fullwall 

Interfacing out the Vanilla AI is simply that...it is meant to be a thin layer. Granted we also want to make some quality of life improvements here and there but ultimately I don't want to shy away from what it is.

Keep in mind that if you were to port Citizens (as an example), you would likely not make use of these. Granted you could (besides AI needing a Wolf for example) but I'd envision you to use your own tasks and pathing. It is my intention to interface out the Vanilla pathfinder as well as include the possibility to create your own. This is the high-level abstraction you are looking for.

**Comment by liach** [comment]
@Zidane: I met an issue when trying to implement few tasks with generics. When I used generics, I can no longer register [these two tasks](https://github.com/SpongePowered/SpongeCommon/blob/master/src/main/java/org/spongepowered/common/registry/type/entity/AITaskTypeModule.java#L101-L102). I have not yet find a solution for that, and, in my opinion, I might make further adjustments on those interfaced AI tasks.

**Comment by Zidane** [comment]
@liach 

Those are registered currently and functional. What is the specific issue you are having?

**Comment by liach** [comment]
Type casting... Intellij Idea says that Classes like `WatchClosestAITask` cannot be cast to `AITask<? extends Agent>`. For other classes that do not include generics, they work well still.

**Comment by fullwall** [comment]
@Zidane - yes no intention of using this PR for Citizens but just curious on what level of behaviour abstraction would be expected for this PR (e.g. setting selection strategy vs setting a filter for the default black-box selection strategy).

File: src/main/java/org/spongepowered/api/entity/ai/task/AITaskTypes.java [hunk 11]
```diff
@@ -25,37 +25,317 @@
...
+import org.spongepowered.api.entity.living.monster.ZombiePigman;
+import org.spongepowered.api.entity.living.Humanoid;
+import org.spongepowered.api.item.ItemType;
+import org.spongepowered.api.item.inventory.ItemStack;
 
 import java.util.function.Predicate;
 
 public final class AITaskTypes {
 
+    // SORTFIELDS:ON
```
Comment (phroa) [inline src/main/java/org/spongepowered/api/entity/ai/task/AITaskTypes.java]: ?

Comment (liach) [inline src/main/java/org/spongepowered/api/entity/ai/task/AITaskTypes.java]: Well I was using SpongeGradle utility before, and now I've removed these.

**Comment by liach** [comment]
@Zidane Do you have any more requests? I've been [trying to implement some tasks](https://github.com/liachmodded/SpongeCommon/tree/feature/more-ais) though.

**Comment by liach** [comment]
Moreover, I have question on AIs from other mods. For example, there's an ai which an `EntityLiving` eats grass in vanilla, and a mod makes an ai subclass which a custom entity from that mod eats grass and explode. Since the type on the interface of the vanilla ai is fixed, it will be a pain to make an interface for that custom ai subclass, if the mod maker want to provide an extra api for sponge plugins.

Few solutions:
- Make the generics in forms of `SomeVanillaAITask<O extends Living>` instead of `SomeVanillaAITask<Living>`; (I might also change it to `SomeVanillaAITask<O extends Living, A extends SomeVanillaAITask<O, A>>` for chaining)
- Let them make ai interfaces which directly extends `AITask<O extends Agent>`; (if this suggestion gets applied, I might change the inheritances of current interfaces to make them not strictly follow vanilla inheritances, for example, `WatchClosestAsInteractingAITask<O extends Agent, A extends WatchClosestAsInteractingAITask<O, A>>` might get removed, and we will just use `WatchClosestAITask<O extends Agent, A extends WatchClosestAITask<O, A>>`)

**Comment by liach** [comment]
@Zidane Updates?

File: src/main/java/org/spongepowered/api/entity/ai/task/AbstractAITask.java [hunk 12]
```diff
@@ -36,13 +36,16 @@
...
- * it instead.
+ * Minecraft, use {@link GameRegistry#createBuilder(Class)} and pass a builder
+ * to it instead.
  *
  * @param <O> The type of Agent
  */
 public abstract class AbstractAITask<O extends Agent> implements AITask<O> {
 
+    /**
+     * The wrapped {@link AITaskType}.
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/entity/ai/task/AbstractAITask.java]: It's not really the 'wrapped' type - just the type of the task.

File: src/main/java/org/spongepowered/api/entity/ai/task/AbstractAITask.java [hunk 13]
```diff
@@ -61,13 +64,33 @@ public final AITaskType getType() {
         return Optional.empty();
     }
 
+    /**
+     * Start the execution of the task.
+     */
     public abstract void start();
 
+    /**
+     * Tells that whether the task can update after starting.
```
Comment (Aaron1011) [inline src/main/java/org/spongepowered/api/entity/ai/task/AbstractAITask.java]: This could just be 'Returns whether the task can update after being started'

**Comment by Aaron1011** [comment]
@liach: Will you have time to work on the implementation for this? If not, some of us on the sponge team may take over finishing this up, in order to get it merged.

**Comment by liach** [comment]
@Aaron1011 Should I make most tasks generic for easy extensions of AIs from other mods? When the generics are erased at runtime it should be working just like the old way, but I need to add some suppress warnings for builders (and some type casts), because the builder does not support type tokens but only classes.

**Comment by gabizou** [comment]
> @Aaron1011 Should I make most tasks generic for easy extensions of AIs from other mods? When the generics are erased at runtime it should be working just like the old way, but I need to add some suppress warnings for builders (and some type casts), because the builder does not support type tokens but only classes.

I honestly would consider writing some of the implementation of this PR first before considering possible additions/expansions for the tasks.

**Comment by liach** [comment]
Example for thaumcraft:

Original AI class:
```java
public class AINearestAttackableTarget
  extends EntityAITarget
{
  EntityGolemBase theGolem;
  /* more stuff not shown*/
}
```

If we interface this in ThaumicSpongeAPI:
(I will add another generic `O` for owner type)
```java
public interface GolemNearastAttackableTargetAITask<O extends Golem, A extends GolemNearastAttackableTargetAITask<O, A>> extends TargetAITask<O, A> {
  //stuff
}
```

File: src/main/java/org/spongepowered/api/entity/ai/Goal.java [hunk 14]
```diff
@@ -64,15 +64,15 @@
      * @param task The task to run
      * @return This goal, for chaining
      */
-    Goal<O> addTask(int priority, AITask<? extends O> task);
+    Goal<O> addTask(int priority, AITask<? super O> task);
```
Comment (liach) [inline src/main/java/org/spongepowered/api/entity/ai/Goal.java]: @Zidane I think you handled these generics incorrectly. Any thoughts on these? Also pinging @kashike to pick out issues.

File: src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/wolf/package-info.java [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+@org.spongepowered.api.util.annotation.NonnullByDefault package org.spongepowered.api.entity.ai.task.builtin.creature.wolf;
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/entity/ai/task/builtin/creature/wolf/package-info.java]: Would be nice to reduce line length, applies to all `package-info`:
```java
@org.spongepowered.api.util.annotation.NonnullByDefault
package org.spongepowered.api.entity.ai.task.builtin.creature.wolf;
```

File: src/main/java/org/spongepowered/api/entity/ai/task/AITaskTypes.java [hunk 16]
```diff
@@ -38,32 +53,291 @@
     // SORTFIELDS:ON
 
     /**
-     * {@link AITask} where {@link Creature}s avoid other {@link Agent}s based on a {@link Predicate}.
+     * {@link AITask} where {@link Tameable}s attack a {@link Living} which its
+     * {@link Tamer} attacks.
+     */
+    public static final AITaskType ATTACK_AFTER_OWNER = DummyObjectProvider.createFor(AITaskType.class, "ATTACK_AFTER_OWNER");
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/entity/ai/task/AITaskTypes.java]: Might be nice to create a `private` method in this class just to reduce "boilerplate":
```java
private static AITaskType createDummyFor(String fieldName) {
    return DummyObjectProvider.createFor(AITaskType.class, fieldName);
}
```

File: src/main/java/org/spongepowered/api/entity/ai/task/builtin/FindNearestEntityAITask.java [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+     *
+     * <p>Note that you still need to set the target class for the task in order
+     * to make the correct entity types included.</p>
+     *
+     * @param filter The predicate of living
+     * @return The task for chaining
+     */
+    FindNearestEntityAITask setFilter(Predicate<Living> filter);
+
+
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/entity/ai/task/builtin/FindNearestEntityAITask.java]: Unnecessary extra newline

File: src/main/java/org/spongepowered/api/entity/ai/task/builtin/OcelotStyleAttackAITask.java [hunk 18]
```diff
@@ -0,0 +1,103 @@
...
+     * @return The task for chaining
+     */
+    OcelotStyleAttackAITask setMaxRange(double maxRange);
+
+    /**
+     * Gets the ticks between two attacks. Default to 20 ticks.
+     *
+     * @return The ticks between two attacks
+     */
+    int getDefaultAttackCountDown();
```
Comment (kashike) [inline src/main/java/org/spongepowered/api/entity/ai/task/builtin/OcelotStyleAttackAITask.java]: Would "cooldown" be a better term?

File: src/main/java/org/spongepowered/api/entity/ai/Goal.java (hunk 14 above)
Comment (ryantheleach) [inline src/main/java/org/spongepowered/api/entity/ai/Goal.java]: So you are passing in an AITask of unknown type, that may contain an O, (or another type) and turning that into a Goal<O>

That sounds like a casting error, an AITask of AITask<Object> might not even contain an O then you attempt to return a Goal<O> ?

Passing in an AITask<? extends O> means that you are guaranteed to have some sort of O inside AITask, you just don't know what the type is. Assuming that addTask doesn't need to know the type, returning Goal<O> is correct.

Comment (liach) [inline src/main/java/org/spongepowered/api/entity/ai/Goal.java]: A task appliable to a `Zombie` may not be applicable to `ZombiePigman`. In `AITask` interface definition, the `<O>` already extends `Agent`.

File: src/main/java/org/spongepowered/api/entity/ai/task/AITaskTypes.java [hunk 19]
```diff
@@ -38,32 +52,295 @@
...
+     */
+    public static final AITaskType TRADE_PLAYER = createDummyFor("TRADE_PLAYER");
+
+    /**
+     * {@link AITask} where {@link Tameable}s attack a target until getting
+     * tamed.
+     */
+    public static final AITaskType UNTAMED_FIND_NEAREST_ATTACKABLE = DummyObjectProvider.createFor(AITaskType.class,
+
+            "UNTAMED_FIND_NEAREST_ATTACKABLE");
```
Comment (ST-DDT) [inline src/main/java/org/spongepowered/api/entity/ai/task/AITaskTypes.java]: This looks like a formatting error.

Comment (Cybermaxke) [inline src/main/java/org/spongepowered/api/entity/ai/task/AITaskTypes.java]: I think that it's a issue with the `sortClassFields` gradle task.

**Comment by gabizou** [comment]
Unless there's more verifiable activity in both merging this PR and implementing this PR, this will be expected for the release after API 5.0.0, be it a content release or the next major API release.

**Comment by Zidane** [comment]
Closing this as it is WAYYYY to big of a PR (for a non core team member) and needs to be broken up into manageable bits. Those bits need implementation code before proceeding.

If you are willing to do that @liach I await your PRs.
